File: test/aggregate.test.ts
import { describe, test, expect } from "bun:test";
import { validatePipeline } from "../src/mongo/aggregate.ts";

describe("validatePipeline", () => {
  test("rejects $out stage", () => {
    const pipeline = [{ $match: { status: "active" } }, { $out: "outputCollection" }];
    expect(() => validatePipeline(pipeline)).toThrow(
      'Write stage "$out" is not allowed. agent-mongo is read-only.',
    );
  });

  test("rejects $merge stage", () => {
    const pipeline = [
      { $group: { _id: "$type", count: { $sum: 1 } } },
      { $merge: { into: "results" } },
    ];
    expect(() => validatePipeline(pipeline)).toThrow(
      'Write stage "$merge" is not allowed. agent-mongo is read-only.',
    );
  });

  test("rejects $out even as first stage", () => {
    const pipeline = [{ $out: "test" }];
    expect(() => validatePipeline(pipeline)).toThrow('Write stage "$out"');
  });

  test("rejects $merge even as first stage", () => {
    const pipeline = [{ $merge: "test" }];
    expect(() => validatePipeline(pipeline)).toThrow('Write stage "$merge"');
  });

  test("allows valid read-only pipeline", () => {
    const pipeline = [
      { $match: { status: "active" } },
      { $group: { _id: "$category", total: { $sum: "$amount" } } },
      { $sort: { total: -1 } },
      { $limit: 10 },
    ];
    expect(() => validatePipeline(pipeline)).not.toThrow();
  });

  test("allows empty pipeline", () => {
    expect(() => validatePipeline([])).not.toThrow();
  });

  test("allows $project, $unwind, $lookup stages", () => {
    const pipeline = [
      { $project: { name: 1, status: 1 } },
      { $unwind: "$items" },
      { $lookup: { from: "other", localField: "id", foreignField: "ref", as: "joined" } },
    ];
    expect(() => validatePipeline(pipeline)).not.toThrow();
  });

  test("skips non-object stages", () => {
    // validatePipeline continues past non-object entries
    const pipeline: unknown[] = [null, undefined, "invalid", { $match: { x: 1 } }];
    expect(() => validatePipeline(pipeline as Record<string, unknown>[])).not.toThrow();
  });
});

File: internal/mongo/aggregate_test.go
package mongo

import (
	"strings"
	"testing"

	"go.mongodb.org/mongo-driver/v2/bson"
)

func TestValidatePipelineRejectsOut(t *testing.T) {
	pipeline := bson.A{
		bson.D{{Key: "$match", Value: bson.D{{Key: "status", Value: "active"}}}},
		bson.D{{Key: "$out", Value: "outputCollection"}},
	}
	err := ValidatePipeline(pipeline)
	if err == nil || err.Error() != `Write stage "$out" is not allowed. agent-mongo is read-only.` {
		t.Errorf("error = %v, want exact $out rejection", err)
	}
}

func TestValidatePipelineRejectsMerge(t *testing.T) {
	pipeline := bson.A{
		bson.D{{Key: "$group", Value: bson.D{{Key: "_id", Value: "$type"}}}},
		bson.D{{Key: "$merge", Value: bson.D{{Key: "into", Value: "results"}}}},
	}
	err := ValidatePipeline(pipeline)
	if err == nil || err.Error() != `Write stage "$merge" is not allowed. agent-mongo is read-only.` {
		t.Errorf("error = %v, want exact $merge rejection", err)
	}
}

func TestValidatePipelineRejectsOutAsFirstStage(t *testing.T) {
	err := ValidatePipeline(bson.A{bson.D{{Key: "$out", Value: "test"}}})
	if err == nil || !strings.Contains(err.Error(), `Write stage "$out"`) {
		t.Errorf("error = %v, want $out rejection", err)
	}
}

func TestValidatePipelineRejectsMergeAsFirstStage(t *testing.T) {
	err := ValidatePipeline(bson.A{bson.D{{Key: "$merge", Value: "test"}}})
	if err == nil || !strings.Contains(err.Error(), `Write stage "$merge"`) {
		t.Errorf("error = %v, want $merge rejection", err)
	}
}

func TestValidatePipelineAllowsReadOnlyPipeline(t *testing.T) {
	pipeline := bson.A{
		bson.D{{Key: "$match", Value: bson.D{{Key: "status", Value: "active"}}}},
		bson.D{{Key: "$group", Value: bson.D{{Key: "_id", Value: "$category"}}}},
		bson.D{{Key: "$sort", Value: bson.D{{Key: "total", Value: -1}}}},
		bson.D{{Key: "$limit", Value: 10}},
	}
	if err := ValidatePipeline(pipeline); err != nil {
		t.Errorf("ValidatePipeline() = %v, want nil", err)
	}
}

func TestValidatePipelineAllowsEmptyPipeline(t *testing.T) {
	if err := ValidatePipeline(bson.A{}); err != nil {
		t.Errorf("ValidatePipeline([]) = %v, want nil", err)
	}
}

func TestValidatePipelineAllowsProjectUnwindLookup(t *testing.T) {
	pipeline := bson.A{
		bson.D{{Key: "$project", Value: bson.D{{Key: "name", Value: 1}}}},
		bson.D{{Key: "$unwind", Value: "$items"}},
		bson.D{{Key: "$lookup", Value: bson.D{{Key: "from", Value: "other"}}}},
	}
	if err := ValidatePipeline(pipeline); err != nil {
		t.Errorf("ValidatePipeline() = %v, want nil", err)
	}
}

func TestValidatePipelineSkipsNonObjectStages(t *testing.T) {
	pipeline := bson.A{nil, "invalid", bson.D{{Key: "$match", Value: bson.D{{Key: "x", Value: 1}}}}}
	if err := ValidatePipeline(pipeline); err != nil {
		t.Errorf("ValidatePipeline() = %v, want nil", err)
	}
}
